### PR TITLE
[SPARK-45746][Python] Return specific error messages if UDTF 'analyze' or 'eval' method accepts or returns wrong values

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This PR adds checks to return specific error messages if any Python UDTF `analyze` or `eval` method accepts or returns wrong values.

Error messages improved include:
* If the `__init__` method takes more arguments than `self` and `analyze_result`.
* If the UDTF call passes more or fewer arguments than `analyze` or `eval` expects (not using `*args` or `**kwargs`).
* If the `analyze` method returns an object besides a `StructType` in the `AnalyzeResult`  `schema` field.
* If there are extra optional `AnalyzeResult` fields relating to input table arguments (e.g. `with_single_partition`) but the `analyze` method received no input table argument.
* If the `analyze` method tries to return a list of strings for the `partition_by` optional field of the `AnalyzeResult` instead of a list of `PartitioningColumn` objects.
* If the `AnalyzeResult` is missing the `schema` argument entirely.
* If we use keyword arguments in the TVF call but the `analyze` or `eval` method does not accept arguments with those keyword(s) (or `**kwargs`).

### Why are the changes needed?

This helps users understand how to easily fix their user-defined table functions if they are malformed.

### Does this PR introduce _any_ user-facing change?

Yes, see above.

### How was this patch tested?

This PR adds test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dtenedor** [comment]
cc @ueshin

**Comment by dtenedor** [comment]
cc @ueshin the failing tests look unrelated

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 1]
```diff
@@ -116,12 +118,150 @@ def main(infile: IO, outfile: IO) -> None:
...
+                and num_expected_args != num_provided_args
+            ):
+                # The UDTF call provided the wrong number of positional arguments.
+                def arguments(num: int) -> str:
+                    return f"{num} argument{'' if num == 1 else 's'}"
+
+                raise PySparkValueError(
+                    format_error(
+                        f"""
+                    {error_prefix} because its '{method}' method expects exactly
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:149]: nit: indent: shall we have the same indent as `f"""` or one more indent?
ditto for the following error messages.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 2]
```diff
@@ -116,12 +118,150 @@ def main(infile: IO, outfile: IO) -> None:
...
+            inspect.getfullargspec(handler.analyze),  # type: ignore[attr-defined]
+            "static analyze",
+            is_static=True,
+        )
+        if hasattr(handler, "eval"):
+            check_method_invariants_before_running(
+                inspect.getfullargspec(handler.eval),  # type: ignore[attr-defined]
+                "eval",
+                is_static=False,
+            )
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:197]: The same/similar check is necessary in `worker.py` for the case without `analyze`?

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 3]
```diff
@@ -116,12 +118,150 @@ def main(infile: IO, outfile: IO) -> None:
...
 
+        error_prefix = f"Failed to evaluate the user-defined table function '{handler.__name__}'"
+
+        def format_error(msg: str) -> str:
+            return dedent(msg).replace("\n", " ")
+
+        # Check invariants about the 'analyze' and 'eval' methods before running them.
+        def check_method_invariants_before_running(
+            expected: inspect.FullArgSpec, method: str, is_static: bool
+        ) -> None:
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:129]: I'm not sure these checks fully cover Python's errors.
I guess we should use `inspect.Signature.bind` and its error message to build new error messages.

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 1 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py:149]: I thought so as well, but the `dev/reformat-python` script keeps changing it back to this :)

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 3 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Good suggestion, this is done.

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 2 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Done.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your review! Responded to your comments, please take another look.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 4]
```diff
@@ -116,12 +118,94 @@ def main(infile: IO, outfile: IO) -> None:
...
+                    f"""
+                    {error_prefix} because the static 'analyze' method expects a result of type
+                    pyspark.sql.udtf.AnalyzeResult with a 'schema' field comprising a StructType,
+                    but the 'schema' field had the wrong type: {type(result.schema)}"""
+                )
+            )
+        has_table_arg = (
+            len([arg for arg in args if arg.isTable])
+            + len([arg for arg in kwargs.items() if arg[-1].isTable])
+        ) > 0
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:168]:
```suggestion
        has_table_arg = any(arg.isTable for arg in args) or any(arg.isTable for arg in kwargs.values())
```

Please reformat it as I'm not sure it follows the format rule.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 5]
```diff
@@ -116,12 +118,94 @@ def main(infile: IO, outfile: IO) -> None:
...
+                    {error_prefix} because the static 'analyze' method returned an
+                    'AnalyzeResult' object with the 'partitionBy' list set to non-empty, but the
+                    function call did not provide any table argument. Please update the query so
+                    that it provides a table argument, or else update the table function so that its
+                    'analyze' method returns an 'AnalyzeResult' object with the 'partitionBy' list
+                    set to empty, and then try the query again."""
+                )
+            )
+        elif (
+            hasattr(result, "partitionBy")
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:194]: Do we need to check this? `AnalyzeResult` should always have this attribute?

**Comment by ueshin** [comment]
cc @allisonwang-db

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 4 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Sounds good, done (also re-ran `dev/reformat-python` afterwards).

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 5 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Good point, removed this check.

**Review by dtenedor — COMMENTED** [review]
Thanks @ueshin for your review! Please take another look.

**Review by ueshin — APPROVED** [review]
LGTM, pending tests.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 6]
```diff
@@ -116,12 +118,89 @@ def main(infile: IO, outfile: IO) -> None:
...
         if not isinstance(result, AnalyzeResult):
             raise PySparkValueError(
-                "Output of `analyze` static method of Python UDTFs expects "
-                f"a pyspark.sql.udtf.AnalyzeResult but got: {type(result)}"
+                format_error(
+                    f"""
+                    {error_prefix} because the static 'analyze' method expects a result of type
+                    pyspark.sql.udtf.AnalyzeResult, but instead this method returned a value of
+                    type: {type(result)}"""
+                )
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py:154]: Good question: these use the error class `TABLE_VALUED_FUNCTION_FAILED_TO_ANALYZE_IN_PYTHON`. Each of these strings goes in the `msg` parameter [1].

[1] https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/python/UserDefinedPythonFunction.scala#L229

File: python/pyspark/worker.py [hunk 7]
```diff
@@ -709,24 +708,28 @@ def read_udtf(pickleSer, infile, eval_type):
     # with one argument containing the previous AnalyzeResult. If that fails, then try a constructor
     # with no arguments. In this way each UDTF class instance can decide if it wants to inspect the
     # AnalyzeResult.
+    udtf_init_args = inspect.getfullargspec(handler)
+
+    udtf_name = handler.__name__
```
[the remark this answers is not in the document]
Comment (dtenedor) [reply python/pyspark/worker.py]: Sounds good, this is done.

**Comment by dtenedor** [comment]
Hi @allisonwang-db, responded to your code review comments, please take another look.

File: python/pyspark/sql/worker/analyze_udtf.py [hunk 8]
```diff
@@ -116,12 +118,89 @@ def main(infile: IO, outfile: IO) -> None:
         handler = read_udtf(infile)
         args, kwargs = read_arguments(infile)
 
+        error_prefix = f"Failed to evaluate the user-defined table function '{handler.__name__}'"
```
Comment (ueshin) [inline python/pyspark/sql/worker/analyze_udtf.py:121]: This `handler.__name__` should also be used the registered name?

**Review by ueshin — APPROVED** [review]
Otherwise, LGTM.

File: python/pyspark/sql/worker/analyze_udtf.py (hunk 8 above)
Comment (dtenedor) [reply python/pyspark/sql/worker/analyze_udtf.py]: Good suggestion, done!

**Comment by ueshin** [comment]
The failed tests seem not related to this PR.

**Comment by ueshin** [comment]
Thanks! merging to master.

**Pull request closed without merging**